### PR TITLE
feat(core): add overflow controller

### PR DESCRIPTION
**Pull request by zeroedin**
## What I did

1.  Added `OverflowController` which when added to a container and given a child array of elements checks to see if those elements exceed the bounds of the container. 
2. Integrated overflow controller into `BaseTabs`.

## Testing Instructions
1. View [deploy preview tabs demo](https://deploy-preview-2375--patternfly-elements.netlify.app/components/tabs/demo/)
2. View at varying browser sizes mobile -> desktop

## Notes to Reviewers

1. Look for any issues around overflow such next and previous buttons not properly displaying the disabled state when tabs are overflowed left or right.
2. Assuming we will want to move the changeset back into the reorganized condensed format.

### CODE REVIEW

File: core/pfe-core/controllers/overflow-controller.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  #scrollTimeoutDelay = 0;
+  #scrollTimeout?: ReturnType<typeof setTimeout>;
+
+  /** Default state */
+  showScrollButtons = false;
+  overflowLeft = false;
+  overflowRight = false;
+
+  /** override to prevent scroll buttons from showing */
+  protected get canShowScrollButtons() {
```
[the remark this answers is not in the document]
Comment (bennypowers) [reply core/pfe-core/controllers/overflow-controller.ts:20]: config prop. camelCase

File: .changeset/eighty-colts-study.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+"@patternfly/pfe-core": patch
```
[the remark this answers is not in the document]
Comment (zeroedin) [reply .changeset/eighty-colts-study.md:2]: @bennypowers do you mind cleaning that up as you see fit?

File: core/pfe-core/controllers/overflow-controller.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeroedin) [reply core/pfe-core/controllers/overflow-controller.ts:20]: That said, the buttons exist in the component render, so the controller can't really enforce if the buttons are present right?  Does this functionality need to be in the controller at all?

[the remark this answers is not in the document]
Comment (nikkimk) [reply core/pfe-core/controllers/overflow-controller.ts:20]: The buttons are really only useful for sighted users using a mouse or touch, so removing them wouldn't cause an accessibility issue.  That said, they do make navigating my mouse and touch easier, so I'm not sure why someone would want to take them away.

[the remark this answers is not in the document]
Comment (nikkimk) [reply core/pfe-core/controllers/overflow-controller.ts:20]: > That said, the buttons exist in the component render, so the controller can't really enforce if the buttons are present right? Does this functionality need to be in the controller at all?

In that case, I don't think it does.

File: elements/pf-tabs/BaseTabs.ts [hunk 3]
```diff
@@ -1,16 +1,19 @@
 import { LitElement, html } from 'lit';
 import { property, query, queryAssignedElements } from 'lit/decorators.js';
+import { classMap } from 'lit/directives/class-map.js';
 
 import { RovingTabindexController } from '@patternfly/pfe-core/controllers/roving-tabindex-controller.js';
-import { getRandomId } from '@patternfly/pfe-core/functions/random.js';
+import { OverflowController } from '@patternfly/pfe-core/controllers/overflow-controller.js';
```
Comment (nikkimk) [inline elements/pf-tabs/BaseTabs.ts:6]: If we are making the overflow buttons optional, should/could we just dynamically import the controller only if the buttons will be used?

**Review by nikkimk — COMMENTED** [review]
Side note: tab and tab panel definitions aren't loading in Safari on https://deploy-preview-2375--patternfly-elements.netlify.app/components/tabs/demo/

File: elements/pf-tabs/BaseTabs.ts (hunk 3 above)
Comment (bennypowers) [reply elements/pf-tabs/BaseTabs.ts:6]: we need to add element-internals-polyfill to the importMap, and we need to make sure to list all our polyfills in our comms to other teams

Comment (bennypowers) [reply elements/pf-tabs/BaseTabs.ts:6]: https://github.com/patternfly/patternfly-elements/pull/2375/commits/b22840d0bb1d7db6a1f98c8a98a1f145e198ae75 should fix this

Comment (bennypowers) [reply elements/pf-tabs/BaseTabs.ts:6]: it's not a bad idea to lazy-load the controller, but that can come afterwards in a later optimization, IMO.

I'd rather work on getting to the point where we can deprecate `BaseTab*` altogether, and just have pf-tab* and rh-tab* implement the same controllers

File: core/pfe-core/controllers/overflow-controller.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bennypowers) [reply core/pfe-core/controllers/overflow-controller.ts:20]: https://lit.dev/docs/composition/controllers/#controllers-and-directives

[the remark this answers is not in the document]
Comment (zeroedin) [reply core/pfe-core/controllers/overflow-controller.ts]: Further abstraction of controller state and render can/will come in a future PR.

File: elements/pf-tabs/BaseTabs.ts [hunk 4]
```diff
@@ -151,26 +149,29 @@ export abstract class BaseTabs extends LitElement {
...
-      <div part="container">
-        <div part="tabs-container">${!this.#showScrollButtons ? '' : html`
+      <div part="container" class="${classMap({ overflow: this.#overflow.showScrollButtons })}">
+        <div part="tabs-container">${!this.#overflow.showScrollButtons ? '' : html`
           <button id="previousTab" tabindex="-1"
               aria-label="${this.getAttribute('label-scroll-left') ?? 'Scroll left'}"
-              ?disabled="${!this.#overflowOnLeft}"
-              @click="${this.#scrollLeft}">
+              ?disabled="${!this.#overflow.overflowLeft}"
+              @click="${this.#scrollLeft.bind(this)}">
```
Comment (bennypowers) [inline elements/pf-tabs/BaseTabs.ts:164]: shouldn't need to explicitly bind instance methods in lit render templates

**Review by bennypowers — CHANGES_REQUESTED** [review]

File: elements/pf-tabs/BaseTabs.ts (hunk 4 above)
Comment (zeroedin) [reply elements/pf-tabs/BaseTabs.ts:164]: Ok weird, before not having it there was causing an issue when `this` in the controller.  Came up while @nikkimk and I were pairing.  Removed.

**Review by bennypowers — APPROVED** [review]

**Pull request merged**